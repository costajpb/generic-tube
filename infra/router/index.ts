import { createRouter, createWebHistory } from 'vue-router'
import DashboardView from '@/infra/views/DashboardView.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'dashboard',
      component: DashboardView
    },
    {
      path: '/shows/:id',
      name: 'show-details',
      component: () => import('@/infra/views/ShowDetailsView.vue')
    }
  ]
})

export default router
