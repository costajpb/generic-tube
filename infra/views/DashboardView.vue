<script setup lang="ts">
    import Dashboard from '@/infra/adapters/use-cases/dashboard'
    import { provide } from 'vue';
    import PageDashboard from '@/infra/components/PageDashboard.vue';
    import type Show from '@/domain/show/entity'
    import router from '@/infra/router';

    const useCase = new Dashboard(['Drama', 'Thriller', 'Crime', 'Adventure','Music'])
    const categories = useCase.categories

    useCase.on('dashboard:display', (show: Show) => {
        router.push(`/shows/${show.id}`)
    })

    provide('useCase', useCase)
</script>

<template>
    <PageDashboard :categories="categories" v-if="categories" />
</template>