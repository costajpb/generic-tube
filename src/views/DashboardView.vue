<script setup lang="ts">
    import Dashboard, { type Categories } from '@/application/dashboard'
    import Shows from '@/domain/show/repository';
    import showAdapter from '@/src/adapters/show-adapter';
    import { onMounted, provide, ref } from 'vue';
    import Emitter from '@/src/util/emitter'
    import MyDashboard from '@/src/components/MyDashboard.vue';
    import type Show from '@/domain/show/entity'
    import router from '@/src/router';

    const container = ref<HTMLElement | null>(null)
    const categories = ref<Categories | undefined>(undefined)
    const display = ref<((show: Show) => void) | null>(null)
    provide('display', display)

    const search = ref<Dashboard['search'] | undefined>(undefined)
    provide('search', search)

    onMounted(async () => {
        const repository = new Shows(showAdapter)
        const useCase = new Dashboard(repository, new Emitter(container.value as HTMLElement))
        categories.value = await useCase.categories
        display.value = useCase.display.bind(useCase)
        search.value = useCase.search.bind(useCase)
        
        container.value?.addEventListener('dashboard:display', (event) => {
            const show = event.detail as Show
            router.push(`/show/${show.id}`)
        })
    })
</script>

<template>
    <div ref="container">
        <MyDashboard :categories="categories" v-if="categories" />
    </div>
</template>