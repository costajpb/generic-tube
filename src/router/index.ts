import { createRouter, createWebHistory } from 'vue-router'
import DashboardView from '@/src/views/DashboardView.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'dashboard',
      component: DashboardView
    },
    {
      path: '/shows/:id',
      name: 'show-details',
      component: () => import('@/src/views/ShowDetailsView.vue')
    }
  ]
})

export default router
