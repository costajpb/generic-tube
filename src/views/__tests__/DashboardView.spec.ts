import { flushPromises, mount } from "@vue/test-utils"
import DashboardView from '../DashboardView.vue'
import router from "@/src/router"
import Dashboard from "@/application/dashboard"
import clickOutside from "@/src/directives/click-outside"

function mountWithDirective(component: any) {
    return mount(component, {
        global: {
            directives: {
                'click-outside': clickOutside
            }
        }
    })
}

describe('DashboardView', () => {
    it('should render properly', () => {
        const wrapper = mountWithDirective(DashboardView)
        expect(wrapper.element).toMatchSnapshot()
    })

    it('should display a show upon request', async () => {
        vi.spyOn(Dashboard.prototype, 'categories', 'get').mockResolvedValue({})
        const push = vi.spyOn(router, 'push')
        const wrapper = mountWithDirective(DashboardView)
        const detail = {id: 1}

        await flushPromises()

        wrapper.trigger('dashboard:display', {
            detail
        })

        expect(push).toHaveBeenCalledWith(`/shows/${detail.id}`)
    })

    it('should allow searching shows with debounce', async () => {
        vi.useFakeTimers()
        const search = vi.spyOn(Dashboard.prototype, 'search').mockImplementation(() => [])
        const wrapper = mountWithDirective(DashboardView)

        await flushPromises()
        
        const input = wrapper.find('[type=search]');

        const query = 'query';

        (input.element as HTMLInputElement).value = query
        
        input.trigger('input')

        vi.runAllTimers()

        expect(search).toHaveBeenCalledWith(query)

        vi.useRealTimers()
    })
})