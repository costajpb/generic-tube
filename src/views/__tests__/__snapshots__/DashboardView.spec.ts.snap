// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardView > should render properly 1`] = `
<div>
  
  <header
    data-v-628430e3=""
  >
    <h1
      class="title"
      data-v-628430e3=""
    >
       ABN Amro Tube 
    </h1>
    <div
      class="extra"
      data-v-628430e3=""
    >
      
      
      <a
        data-v-af51883b=""
        href="#search"
      >
        Search
      </a>
      <section
        class="container"
        data-is-active="false"
        id="search"
      >
        <div
          class="actions"
        >
          <a
            href="#"
          >
            Close
          </a>
          <input
            placeholder="Search shows..."
            type="search"
          />
        </div>
        <!--v-if-->
      </section>
      
      
    </div>
  </header>
  
  
  
</div>
`;
